pkg/utils/file_filter.go

```diff
@@ -0,0 +1,167 @@
+package utils
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	gitignore "github.com/sabhiram/go-gitignore"
+)
+
+type FileFilter struct {
+	path         string
+	defaultRules []string
+}
+
+func NewFileFilter(path string) *FileFilter {
+	return &FileFilter{
+		path:         path,
+		defaultRules: []string{"**/.git/**"},
+	}
+}
+
+// GetAllFiles traverses a given dir path and fetches all filesToFilter in the directory
+func (fw *FileFilter) GetAllFiles() chan string {
+	var filesCh = make(chan string)
+	go func() {
+		defer close(filesCh)
+		err := filepath.WalkDir(fw.path, func(path string, d fs.DirEntry, err error) error {
+			if !d.IsDir() && err == nil {
+				filesCh <- path
+			}
+			return err
+		})
+		if err != nil {
+			fmt.Printf("walk dir failed: %v", err)
+		}
+	}()
+
+	return filesCh
+}
+
+// GetRules builds a list of glob patterns that can be used to filter filesToFilter
+func (fw *FileFilter) GetRules(ruleFiles []string) ([]string, error) {
+	files := fw.GetAllFiles()
+
+	// iterate filesToFilter channel and find ignore filesToFilter
+	var ignoreFiles = make([]string, 0)
+	for file := range files {
+		for _, ruleFile := range ruleFiles {
+			if strings.Contains(file, ruleFile) {
+				ignoreFiles = append(ignoreFiles, file)
+			}
+		}
+	}
+
+	// iterate ignore filesToFilter and extract glob patterns
+	globs, err := fw.buildGlobs(ignoreFiles)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(fw.defaultRules, globs...), nil
+}
+
+// GetFilteredFiles returns a filtered channel of filepaths from a given channel of filespaths and glob patterns to filter on
+func (fw *FileFilter) GetFilteredFiles(filesCh chan string, globs []string) chan string {
+	var filteredFilesCh = make(chan string)
+
+	// create pattern matcher used to match filesToFilter to glob patterns
+	globPatternMatcher := gitignore.CompileIgnoreLines(globs...)
+
+	go func() {
+		defer close(filteredFilesCh)
+		// iterate the filesToFilter channel
+		for file := range filesCh {
+			// filesToFilter that do not match the glob pattern are filtered
+			if !globPatternMatcher.MatchesPath(file) {
+				filteredFilesCh <- file
+			}
+		}
+	}()
+
+	return filteredFilesCh
+}
+
+// buildGlobs iterates a list of ignore filesToFilter and returns a list of glob patterns that can be used to test for ignored filesToFilter
+func (fw *FileFilter) buildGlobs(ignoreFiles []string) ([]string, error) {
```

Comment:
Suggestion: all the functions here and below don't require a receiver as they don't persist a state. Just be clear about this and remove the receivers.